### ISSUE TITLE
Is Kp = 20 too soft? 

### ISSUE THREAD

**Issue by breakds** [Improbable-AI/walk-these-ways#39]
Hi,

First I would like to express my gratitude again on the great work of open sourcing the project and providing the deployment procedure!

I just started trying the deployment of the checkpoint provided in this repo. In the paper Kp = 20 is used as the stiffness for the joint controller and this can be confirmed from both the training code and deployment code.

Based on my experience with unitree Go1 I think Kp = 20 might be too soft. In the deployment script `deployment_runner.py` the `calibrate()` function will create an interpolated action sequence from the current qpos to the target qpos, which is 

`[0.1, 0.8, -1.5, -0.1, 0.8, -1.5, 0.1, 1, -1.5, -0.1, 1, -1.5]`

I  tried the same thing on our Unitree Go1 and it ends up "kneeling" on the ground with qpos below

`[-0.07242385  0.66453123 -1.6734996   0.06848777  0.670284   -1.6667578, -0.13219169  0.8420787  -2.137513    0.12692341  0.8130728  -2.0752623 ]`

Note that the rear leg joints have very big terminal errors.

And pictures:

![IMG_2774](https://github.com/Improbable-AI/walk-these-ways/assets/1111035/2c9b7ad3-61f8-415d-a132-43d589891469)
![IMG_2773](https://github.com/Improbable-AI/walk-these-ways/assets/1111035/15b34616-dae5-46d0-bf1f-e7f9d034b8c4)

Would like to confirm with the author before proceeding - is such behavior expected?

Thanks!

**Comment by gmargo11** [comment]
Hi @breakds ,

Yes, this is the expected behavior. The calibration / stand-up move itself is not tuned much. However, the robot will walk normally when you proceed.

I agree that Kp=20 is pretty low for this robot. I chose this intentionally because it gives the joints some natural compliance. I haven't explored tuning it much. At Kp=20,  the policy still learns to compensate and execute a good walking gait. 

-Gabe

**Comment by breakds** [comment]
> Hi @breakds ,
> 
> Yes, this is the expected behavior. The calibration / stand-up move itself is not tuned much. However, the robot will walk normally when you proceed.
> 
> I agree that Kp=20 is pretty low for this robot. I chose this intentionally because it gives the joints some natural compliance. I haven't explored tuning it much. At Kp=20, the policy still learns to compensate and execute a good walking gait.
> 
> -Gabe

Thanks a lot for the confirmation, Gabe!

**Comment by GuoPingPan** [comment]
Have you ever tried kp=50? Is it to big for these project? I don't know if my go1 was damaged because I hung it up for deploying(the feet off the ground) or because kp was too big.
I train with kp=50 in simulator and performance well.

**Comment by breakds** [comment]
> Have you ever tried kp=50? Is it to big for this project? I don't know if my go1 was damaged because I hung it up for deploying(the feet off the ground) or because kp was too big. I train with kp=50 in simulator and performance well.

I have tried training a simpler policy with Kp = 50 in Mujoco and deploy it with Kp = 50 and it works. The checkpoint provided by Gabe with Kp = 20 also works well when deployed with Kp = 20.

> I train with kp=50 in simulator and performance well.

Do you mean you can successfully deploy a policy trained with `Kp = 50` on your Go1? In that case, why do you suspect your Go1 was damaged?

**Comment by GuoPingPan** [comment]
> > Have you ever tried kp=50? Is it to big for this project? I don't know if my go1 was damaged because I hung it up for deploying(the feet off the ground) or because kp was too big. I train with kp=50 in simulator and performance well.
> 
> I have tried training a simpler policy with Kp = 50 in Mujoco and deploy it with Kp = 50 and it works. The checkpoint provided by Gabe with Kp = 20 also works well when deployed with Kp = 20.
> 
> > I train with kp=50 in simulator and performance well.
> 
> Do you mean you can successfully deploy a policy trained with `Kp = 50` on your Go1? In that case, why do you suspect your Go1 was damaged?

No, I mean it only work well in simulator.
But judging from your experiment, `kp=50` might not be the reason for the deployment failed in my go1.

**Comment by breakds** [comment]
> > > Have you ever tried kp=50? Is it to big for this project? I don't know if my go1 was damaged because I hung it up for deploying(the feet off the ground) or because kp was too big. I train with kp=50 in simulator and performance well.
> > 
> > 
> > I have tried training a simpler policy with Kp = 50 in Mujoco and deploy it with Kp = 50 and it works. The checkpoint provided by Gabe with Kp = 20 also works well when deployed with Kp = 20.
> > > I train with kp=50 in simulator and performance well.
> > 
> > 
> > Do you mean you can successfully deploy a policy trained with `Kp = 50` on your Go1? In that case, why do you suspect your Go1 was damaged?
> 
> No, I mean it only work well in simulator. But judging from your experiment, `kp=50` might not be the reason for the deployment failed in my go1.

I see. What kind of failure did you observe?

**Comment by GuoPingPan** [comment]
All the motors had been demanged. I already sent it to be repaired.
I hung up go1 without feet contacted with the ground while deploying. Is it the reason? Have you ever try to deploy while hanging up go1 without feet contacted with the ground?

**Comment by breakds** [comment]
There should be power protection and we set it to 9. Did you do the same?

Also, we always deploy with feet on the ground, sometimes with a rope loosely hanging it

**Comment by GuoPingPan** [comment]
I use the same power protection level too. Maybe I should deploy as you say. Thanks a lot.

**Comment by GuoPingPan** [comment]
> > Have you ever tried kp=50? Is it to big for this project? I don't know if my go1 was damaged because I hung it up for deploying(the feet off the ground) or because kp was too big. I train with kp=50 in simulator and performance well.
> 
> I have tried training a simpler policy with Kp = 50 in Mujoco and deploy it with Kp = 50 and it works. The checkpoint provided by Gabe with Kp = 20 also works well when deployed with Kp = 20.
> 
> > I train with kp=50 in simulator and performance well.
> 
> Do you mean you can successfully deploy a policy trained with `Kp = 50` on your Go1? In that case, why do you suspect your Go1 was damaged?

@breakds 
Sorry for bother you again. You mean you just used `kp=50` in mujoco for a simple policy but not walk-these-ways? Have you ever tried to retrain a model and deploy for walk-these-ways? Though the pretrained-model perform well, I found that the reward of the self-retrained model with only modifing `control_type='P'` can't coverge. Just as bellow:

> Is there anything wrong with my behaviours?

![image](https://github.com/Improbable-AI/walk-these-ways/assets/69315544/13b9f4c0-2998-44a5-ac35-1170726c1871)
